keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
+  name: MY-BUCKET-1 [1]
+  namespace: OBC-NAMESPACE [2]
+  finalizers: [3]
+  - objectbucket.io/finalizer
+  labels: [4]
+    bucket-provisioner: aws-s3.io-bucket [5]
+  ownerReferences: [6]
+  - name: MY-BUCKET-1
+    ...
+data: 
+  BUCKET_HOST: http://MY-STORE-URL [7]
+  BUCKET_PORT: 80 [8]
+  BUCKET_NAME: MY-BUCKET-1 [9]
+  BUCKET_REGION: us-west-1
+  ... [10]
+```
+1. same name as the OBC. Unique since the configMap is in the same namespace as the OBC.
+1. determined by the namespace of the ObjectBucketClaim.
+1. finalizers set and cleared by the lib's OBC controller. Prevents accidental deletion of the ConfigMap.
+1. the library adds a label (seen here) but each provisioner can supply their own labels.
+1. the label value shown is the name of the provisioner but due to Kubernetes restrictions slash (/) is
+   replaced by a dash (-). In this example the provisioner name is `aws-s3.io/bucket`.
+1. ownerReference sets the ConfigMap as a child of the ObjectBucketClaim. Deletion of the ObjectBucketClaim causes the deletion of the ConfigMap.
+1. host URL.
+1. host port.
+1. unique bucket name.
+1. the above data keys are defined by the library.
+Provisioners are able to cause the lib to create additional data keys by returning the `AdditionalConfigData` field.
+
+#### Storage Class (created by admin)
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: MyAwsS3Class
+  labels: 
+    aws-s3/object [1]
+provisioner: aws-s3.io/bucket [2]
+parameters: [3]
+  region: us-west-1
+  secretName: s3-bucket-owner
+  secretNamespace: s3-provisioner
+  bucketName: existing-bucket [4]
+reclaimPolicy: Delete [5]
+```
+1. (optional) the label here associates this StorageClass to a specific provisioner.
+1. provisioner responsible for handling OBCs referencing this StorageClass.
+1. **all** parameter keys and values are specific to a provisioner, are optional, and are not validated by the StorageClass API.
+Fields to consider are object-store endpoint, version, possibly a secretRef containing info about credential for new bucket owners, etc.
+1. bucketName is required for access to existing buckets.
+Unlike greenfield provisioning, the brownfield bucket name appears in the storage class, not the OBC.
+1. each provisioner decides how to treat the _reclaimPolicy_ when an OBC is deleted. Supported values are:
++ _Delete_ = (typically) physically delete the bucket.
+Depending on new vs. existing bucket, the provisioner's `Delete` or `Revoke` methods are called.
++ _Retain_ = (typically) do not physically delete the bucket.
+Depending on the provisioner, various clean up steps can be performed, such as deleting users, revoking credentials, etc.
+For both new and existing buckets the provisioner's `Revoke` method is called.
+
+#### App Pod (independent of provisioner)
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: app-pod
+  namespace: dev-user
+spec:
+  containers:
+  - name: mycontainer
+    image: redis
+    envFrom: [1]
```

Comment:
This is just an example, right?  Mounted secrets might be preferred (e.g., in the case of multiple bucket claims)

Reply:
Yes.  It would probably be helpful to add that case as well.